rfc/0076-subdoc-replica-reads.md

```diff
@@ -0,0 +1,179 @@
+# Meta
+
+| Field          | Value                      |
+|----------------|----------------------------|
+| RFC Name       | Sub-Document Replica Reads |
+| RFC ID         | 76                         |
+| Start Date     | 2023-06-20                 |
+| Owner          | Charles Dixon              |
+| Current Status | DRAFT                      |
+| Revision       | #1                         |
+
+# Summary
+[MB-23162](https://issues.couchbase.com/browse/MB-23162) adds support for performing subdocument read operations on replica servers.
+Replica read operations are performed against replica servers with a new ReplicaRead protocol flag.
+it also includes a new, informational, HELLO flag to be negotiated.
+
+This RFC specifies the SDK work needed to support this feature.
+All SDKs are required to implement this feature.
+
+# Motivation
+SDKs require a way to expose this new ReplicaRead feature in a way that is idiomatic to the SDK API.
+
+# Changes
+
+1. Add two new functions to the `Collection` interface - `LookupInAnyReplica` and `LookupInAllReplicas`.
+2. Add a new `LookupInReplicaResult` result type.
+3. Add two new options types - `LookupInAnyReplicaOptions` and `LookupInAllReplicasOptions`.
+4. Support the new REPLICA_READ (0x1c) HELLO flag.
+5. When the above functions are used send `LookupIn` requests to replica nodes specifying the REPLICA_READ protocol flag (0x20).
+
+## LookupInAllReplicas
+
+Gets a stream of document data from the server using LookupIn, leveraging both the active and all available replicas; the caller can choose which replica they wish to take and cancel the stream at any time.
+This MUST fetch from both the active and all replicas due to failover situations that can arise.
+In addition, the SDK MUST broadcast the requests simultaneously to ensure the application receives data as quickly as possible.
+
+```
+Stream<ILookupInReplicaResult> LookupInAllReplicas(string id, LookupInSpec[] specs, [LookupInAllReplicasOptions])
+```
+
+Parameters
+
+- Required
+
+    - Id: string - the primary key.
+    - LookupInSpec - an array of fetch operations, see [LookupIn](./0053-sdk3-crud.md#LookupIn)
+
+- Optional
+    - Timeout or timeoutMillis (int/duration) - the time allowed for the operation to be terminated. This is controlled by the client.
+    - Serializer - a custom serializer for converting the memcached JSON data to a native type.
```

Comment:
Plus the standard parentSpan().

What we'd do with RetryStrategy is less clear, but is somewhat tricky to not support in the JVM clients as it's part of CommonOptions.

Reply:
Those are both missing because they aren't in the original SDK 3 crud api RFC that I copied from. Will add.